README.md
# Sessioner

Creates and maintains specific number of TCP sessions between two hosts. The number of sessions can be dynamically adjusted when it is running.

## Usage 

Sessioner consists of two binaries - `sessioner_initiator` and `sessioner_responder`. 
- Responder show be started first because it listens on configured TCP ports and accepts the connections from initiator.
- Initiator then establishes the configured number of connections to responder and maintains them.

After started, each program shows its own prompt for getting runtime status and changing the number of required connections.

Impatients can jump directly to [Example section](#example).

## sessioner_responder

### Parameters

```
Usage: ./sessioner_responder [-dh] --listen ip:port[-port]
       -h         ... show this help
       -d         ... enable debug outputs
       -u         ... do not try to change resource limits on file descriptors
       --listen   ... listen on IP address and port
                      optionally port range can be specified
                      this option can appear multiple times
       --hashsize ... size of primary hashing table for clients
```

- `-d` Enable debug outputs. This can generate a lot of data for high number of connections.
- `-u` Normally the program tries to read the system limit of maximum open file descriptor per process and sets this limit. This option disables it. Be aware
       that default limit of FD is usually 1024 which is not enough for any serious testing.
-  `--hashsize` Size of primary hashing table (which is used for all clients from all listeners), default is 10,000 which is enough for about 100,000 connected
                clients. When you are testing with more clients you will need to increase this appropriately otherwise accepting new clients will be quite slow.
- `--listen` The most important and the only required parameter. It specified where the responder should wait for new connections. It can be specified many times.
   It can be as simple as `10.0.0.1:5000` or it can be a continous range of ports to listen on like `10.0.0.1:5000-5999`. Reason for using multiple listening
   ports is that for two hosts and one listening port there are usually only about 28,000 possible connections (see [Ports selection](#port-selection) section).
   
### Commands

Reponder's prompt is `responder $ ` and you can type `help` to list available commands. Following are currently recognized:

- `status` Shows some statistics counter. Most important is `Connected clients` which shows the number of connected clients from responder's point of view.
```
responder $ status
Connected clients        : 15
Active listeners         : 1
Clients hash size        : 100000
Clients hash max depth   : 1
Listeners hash max depth : 1
Max descriptors limit    : 1048576/1048576
Debug output             : no
```

- `list` Shows basic IP information about every client.
```
responder $ list
Connected clients according to internal hash structure: 15
127.0.0.1:49496->127.0.0.1:6000 : last action 107 seconds ago
127.0.0.1:49498->127.0.0.1:6000 : last action 93 seconds ago
127.0.0.1:49500->127.0.0.1:6000 : last action 93 seconds ago
127.0.0.1:49502->127.0.0.1:6000 : last action 93 seconds ago
127.0.0.1:49504->127.0.0.1:6000 : last action 93 seconds ago
127.0.0.1:49506->127.0.0.1:6000 : last action 93 seconds ago
127.0.0.1:49508->127.0.0.1:6000 : last action 93 seconds ago
127.0.0.1:49510->127.0.0.1:6000 : last action 93 seconds ago
127.0.0.1:49512->127.0.0.1:6000 : last action 93 seconds ago
127.0.0.1:49514->127.0.0.1:6000 : last action 93 seconds ago
127.0.0.1:49516->127.0.0.1:6000 : last action 60 seconds ago
127.0.0.1:49518->127.0.0.1:6000 : last action 60 seconds ago
127.0.0.1:49520->127.0.0.1:6000 : last action 60 seconds ago
127.0.0.1:49522->127.0.0.1:6000 : last action 60 seconds ago
127.0.0.1:49524->127.0.0.1:6000 : last action 60 seconds ago
```

- `listeners` Shows basic information about every configured listener, including the number of clients it has accepted since start.
```
responder $ listeners
Active listeners according to internal hash structure: 1
127.0.0.1:6000 : 28247 accepted connections
```

- `debug` Enable or disable printing the debug output.
```
responder $ debug
Debug output enabed
responder $ debug
Debug output disabled
```

- `quit` Terminate all the connections and the responder.

Pressing enter without typing any command will just repeat the last command.

## sessioner_initiator

### Parameters

```
Usage: ./sessioner_initiator [-dph] --target ip:port[-port] [--hashsize NUM]
       -h         ... show this help
       -d         ... enable debug outputs
       -p         ... ping through connections periodically
       -u         ... do not try to change resource limits on file descriptors
       --target   ... connect on IP address and port
                      optionally port range can be specified
                      this option can appear multiple times
       --hashsize ... size of primary hashing table for following targets
```

- `-d` Enable debug outputs. This can generate a lot of data for high number of connections.
- `-u` Normally the program tries to read the system limit of maximum open file descriptor per process and sets this limit. This option disables it. Be aware
       that default limit of FD is usually 1024 which is not enough for any serious testing.
- `-p` Enable sending an explicit "PING" packet though the session after some period of time. Currently the ping is sent every 30-60 seconds (randomized for
       every session).
- `--hashsize` Similar to the same parameter in responder, but in initiator the session table is independent for every target (one IP and and port). Every
  `--target` configured after `--hashsize` parameters will have the specific size configured. It is usually not necessary to change the default (10,000).
- `--target` Specifies the remote IP and port where responder is listening. It can be specified many times. It can be as simple as `10.0.0.1:5000` or it 
             can be a continous range of ports to listen on like `10.0.0.1:5000-5999`. Reason for using multiple listening ports is that for two hosts and
             one listening port there are usually only about 28,000 possible connections (see [Ports selection](#port-selection) section).

### Commands

Initiator's prompt is `initiator $ ` and you can type `help` to list available commands. Following are currently recognized:

- `status` Shows some statistics counter. Most important is `Connections requested` which shows the desired total number of sessions and `Connected clients`
  which shown the actual number of established sessions. It can take some time for all the requested sessions to be established, but these two counters
  should eventualy match.
```
initiator $ status
Connections requested    : 40
Connected clients        : 40
Configured targets       : 4
Max descriptors limit    : 1048576/1048576
Ping through connections : no
Debug output             : no
```

- `details` Shows details about every target (single IP and port). Notice that the total number of clients is 40 and it is eqully distributed between
  all the targets.
```
initiator $ details
Target 127.0.0.1:6000:
   Maximum allowed connections       : 10
   Currently established connections : 10
   Session hash max depth            : 1
Target 127.0.0.1:6001:
   Maximum allowed connections       : 10
   Currently established connections : 10
   Session hash max depth            : 1
Target 127.0.0.1:6002:
   Maximum allowed connections       : 10
   Currently established connections : 10
   Session hash max depth            : 1
Target 127.0.0.1:6003:
   Maximum allowed connections       : 10
   Currently established connections : 10
   Session hash max depth            : 1
```

- `set NUM` Sets the total number of sessions to keep established. The sessions are automatically distributed between all targets. If the new limit
  is lower than the previous one, some sessions are automatically closed.

```
initiator $ set 60
```

- `ping` Enable or disable sending explicit "PING" packet though every sessions from time to time. Currently the ping is sent every 30-60 seconds
  (randomized for every session).
```
initiator $ ping
Ping enabed
initiator $ ping
Ping disabled
```

- `debug` Toggle the debug output printing.
```
initiator $ debug
Debug output enabed
initiator $ debug
Debug output disabled
```

- `quit` Terminate all sessions and quit.

Pressing enter without typing any command will just repeat the last command.


## Other notes

### Port selection

When testing is happening between two hosts with one usable IP address on each and you specify ony one listening port, the maximal number of possible
sessions that can be established on generic Linux distribution is about 28,000. It is because of following sysctl setting:

```
# sysctl net.ipv4.ip_local_port_range
net.ipv4.ip_local_port_range = 32768	60999
```

This sysctl specifies the ports that can be used for outgoing connections. In the case there are 28,231 ports available. 

You can change this sysctl, but it is better to specify more ports and/or IP addresses. Then every combination of IP and port can establish about 28,000
sessions. For 1 million sessions you need about 40 combinations. 

### Maximum number of file descriptors

Both initiator and responder will try to configure the highest possible number of open file descriptors, unless `-u` parameter was specified.

This limit is specified in `/proc/sys/fs/nr_open` and is usually slighly above 1 million. 
```
# cat /proc/sys/fs/nr_open
1048576
```

That is why it is not possible to open more than roughly 1 million sessions between one initiator and responder.

## Example 

- Start responder on host testB. Listen on single IP address and one hundred ports from 3000 to 3099. In the status output we can see there are
  indeed 100 configured listeners and the are no connected clients yet.
```
root@testB:~# sessioner_responder --hashsize 100000 --listen 10.109.80.40:3000-3099
responder $ status
Connected clients        : 0
Active listeners         : 100
Clients hash size        : 100000
Clients hash max depth   : 0
Listeners hash max depth : 1
Max descriptors limit    : 1048576/1048576
Debug output             : no
```

- Start initiator on hostA. Configure 100 targets - all on IP 10.109.80.40 and ports from 3000 to 3099. After start the initiator establishes
  one session per target, therefore there is now 100 sessions established.
```
root@testA:~# sessioner_initiator --target 10.109.80.40:3000-3099
initiator $ status
Connections requested    : 100
Connected clients        : 100
Configured targets       : 100
Max descriptors limit    : 1048576/1048576
Ping through connections : no
Debug output             : no
```

- We can verify the same numbers of established sessions are seen on responder.
```
responder $ status
Connected clients        : 100
Active listeners         : 100
Clients hash size        : 100000
Clients hash max depth   : 1
Listeners hash max depth : 1
Max descriptors limit    : 1048576/1048576
Debug output             : no
```

- Now configure initiator to have 100,000 sessions established.
```
initiator $ set 100000
```

- It will take some time. We can monitor the progress by executing `status` command. Empty command will repeat the previous command.
```
initiator $ status
Connections requested    : 100000
Connected clients        : 2652
Configured targets       : 100
Max descriptors limit    : 1048576/1048576
Ping through connections : no
Debug output             : no
initiator $
Connections requested    : 100000
Connected clients        : 33900
Configured targets       : 100
Max descriptors limit    : 1048576/1048576
Ping through connections : no
Debug output             : no
initiator $
Connections requested    : 100000
Connected clients        : 63036
Configured targets       : 100
Max descriptors limit    : 1048576/1048576
Ping through connections : no
Debug output             : no
initiator $
Connections requested    : 100000
Connected clients        : 100000
Configured targets       : 100
Max descriptors limit    : 1048576/1048576
Ping through connections : no
Debug output             : no
```

- We can do the same monitoring on responder.
```
responder $ status
Connected clients        : 8650
Active listeners         : 100
Clients hash size        : 100000
Clients hash max depth   : 1
Listeners hash max depth : 1
Max descriptors limit    : 1048576/1048576
Debug output             : no
responder $
Connected clients        : 23300
Active listeners         : 100
Clients hash size        : 100000
Clients hash max depth   : 1
Listeners hash max depth : 1
Max descriptors limit    : 1048576/1048576
Debug output             : no
responder $
Connected clients        : 34836
Active listeners         : 100
Clients hash size        : 100000
Clients hash max depth   : 1
Listeners hash max depth : 1
Max descriptors limit    : 1048576/1048576
Debug output             : no
responder $
Connected clients        : 64073
Active listeners         : 100
Clients hash size        : 100000
Clients hash max depth   : 1
Listeners hash max depth : 1
Max descriptors limit    : 1048576/1048576
Debug output             : no
responder $
Connected clients        : 100000
Active listeners         : 100
Clients hash size        : 100000
Clients hash max depth   : 1
Listeners hash max depth : 1
Max descriptors limit    : 1048576/1048576
Debug output             : no
```

- Now we have 100,000 connected sessions.


debian/changelog
sessioner (1.0) unstable; urgency=medium

  * Initial release

 -- Ondrej Holecek <ondrej@holecek.eu>  Tue, 28 Jul 2020 09:11:51 +0200


hash.c
#include <stdio.h>
#include <stdlib.h>
#include "hash.h"

struct intHash *hashInit(unsigned int primaryTableSize) {
	struct intHash *hash = (struct intHash *)malloc(sizeof(struct intHash));
	hash->primaryTableSize = primaryTableSize;
	hash->data             = (struct hashData**)malloc(sizeof(struct hashData *) * primaryTableSize);
	hash->active           = 0;

	for (unsigned int i = 0; i < primaryTableSize; i++) {
		hash->data[i] = NULL;
	}

	pthread_mutex_init(&hash->lock, NULL);

	return hash;
}

void hashDestroy(struct intHash *hash) {
	pthread_mutex_destroy(&hash->lock);
	free(hash->data);
	free(hash);
}

void hashAdd(struct intHash *hash, unsigned int index, void *data) {
	pthread_mutex_lock(&hash->lock);

	unsigned int      hi = index % hash->primaryTableSize;
	struct hashData   *field = hash->data[hi];

	// this section is empty
	if (field == NULL) {
		field = (struct hashData *)malloc(sizeof(struct hashData));
		field->index = index;
		field->value = data;
		field->prev  = NULL;
		field->next  = NULL;
		hash->data[hi] = field;
		hash->active++;
		pthread_mutex_unlock(&hash->lock);
		return;
	}
	
	// non empty, find the last index
	// when going through the list check if the index does not already exist
	struct hashData *prev = field;
	while (field != NULL) {
		if (field->index == index) {
			field->value = data;
			pthread_mutex_unlock(&hash->lock);
			return;
		}

		prev  = field;
		field = field->next;
	}

	// if we got here, there was no previous index
	prev->next = (struct hashData *)malloc(sizeof(struct hashData));
	prev->next->index = index;
	prev->next->value = data;
	prev->next->prev  = prev;
	prev->next->next  = NULL;
	hash->active++;
	pthread_mutex_unlock(&hash->lock);
	return;
}

int hashDelete(struct intHash *hash, unsigned int index) {
	pthread_mutex_lock(&hash->lock);
	
	unsigned int      hi = index % hash->primaryTableSize;
	struct hashData   *field = hash->data[hi];

	if (field == NULL) {
		pthread_mutex_unlock(&hash->lock);
		return 0;
	}

	while (field != NULL) {
		if (field->index == index) {

			if (field->next == NULL && field->prev == NULL) {
				hash->data[hi] = NULL;

			} else if (field->next == NULL && field->prev != NULL) {
				field->prev->next = NULL;

			} else if (field->next != NULL && field->prev == NULL) {
				hash->data[hi] = field->next;
				hash->data[hi]->prev = NULL;

			} else { // field->next != NULL && field->prev != NULL
				// we can never get here
				field->prev->next = field->next;
				field->next->prev = field->prev;
			}

			free(field);
			hash->active--;
			pthread_mutex_unlock(&hash->lock);
			return 1;

		}

		field = field->next;
	}

	fprintf(stderr, "Unable to delete non-existing index %u\n", index);

	pthread_mutex_unlock(&hash->lock);
	return 0;
}

unsigned int hashGetActive(struct intHash *hash) { 
	pthread_mutex_lock(&hash->lock);
	unsigned int active = hash->active;
	pthread_mutex_unlock(&hash->lock);
	return active;
}

unsigned int hashGetDepth(struct intHash *hash) {
	unsigned int maxDepth = 0;

	for (unsigned int i = 0; i < hash->primaryTableSize; i++) {
		struct hashData *field = hash->data[i];
		unsigned int depth = 0;
		
		while (field != NULL) {
			depth++;
			field = field->next;
		}

		if (depth > maxDepth) { maxDepth = depth; }
	}

	return maxDepth;
}

void **hashGetAll(struct intHash *hash, unsigned int *count) {
	pthread_mutex_lock(&hash->lock);

	void **all = (void **)malloc(sizeof(void *) * hash->active);
	*count = 0;

	for (unsigned int i = 0; i < hash->primaryTableSize; i++) {
		struct hashData *field = hash->data[i];
		
		while (field != NULL) {
			all[(*count)++] = field->value;	
			field = field->next;
		}
	}

	pthread_mutex_unlock(&hash->lock);
	return all;
}



client.h
#ifndef __CLIENT_H__
#define __CLIENT_H__

struct direction {
	char            *remoteIp;
	unsigned short  remotePort;
	unsigned int    maxConnections;
	struct intHash  *sessions;
	unsigned int    pingTime;
	unsigned int    pingMax;
};

struct connection {
	int               sockfd;
	struct direction  *direction;
	char              *localIp;
	unsigned short    localPort;
	char              *remoteIp;
	unsigned short    remotePort;
	time_t            connected;
	time_t            lastSent;
	time_t            lastReceived;
	unsigned int      pingRandom;
};

struct direction *directionInit(char *remoteIp, unsigned short remotePort, unsigned int maxConnections);
int createConnection(char *ip, unsigned short port);
void directionCheckup(int ep, struct direction *dir);
void closeConnection(int ep, struct connection *con);
void handleEpoll(int ep);


#endif


server.h
#ifndef __SERVER_H__
#define __SERVER_H__

enum  socketType{LISTENER, CLIENT};

union socketData {
	struct listenerInfo  *listener;
	struct clientInfo    *client;
};

struct socketInfo {
	int              sockfd;
	enum socketType  type;
	time_t           lastAction;
	union socketData data;
};

struct clientInfo {
	char             *localIp;
	unsigned short   localPort;
	char             *remoteIp;
	unsigned short   remotePort;
};

struct listenerInfo {
	char             *listenIp;
	unsigned short   listenPort;
	unsigned int     acceptedConnections;
};

int createListener(unsigned short port, char *ip);
void registerEpollListeners(int ep, char *ip, unsigned short portStart, unsigned short portEnd);
void handleEpoll(int ep);
void handleNewClient(int ep, struct socketInfo *sInfo);
void handleDataFromClient(int ep, struct socketInfo *sInfo);

#endif


client.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <time.h>
#include <getopt.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/resource.h>
#include <sys/epoll.h>
#include "hash.h"
#include "client.h"

int terminate = 0;
int debug     = 0;
int ping      = 0;

unsigned int hashSize = 10000;

struct direction *directionInit(char *remoteIp, unsigned short remotePort, unsigned int maxConnections) {
	struct direction *dir = (struct direction *)malloc(sizeof(struct direction));
	dir->remoteIp       = strdup(remoteIp);
	dir->remotePort     = remotePort;
	dir->maxConnections = maxConnections;
	dir->sessions       = hashInit(hashSize);
	dir->pingTime       = 30;
	dir->pingMax        = 1000;
}

int createConnection(char *ip, unsigned short port) {
	int sockfd = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
	if (sockfd == -1) {
		fprintf(stderr, "Unable to create socket for %s:%u: %s\n", ip, port, strerror(errno));
		return -1;
	}

	struct sockaddr_in saddr;
	saddr.sin_family      = AF_INET;
	saddr.sin_port        = htons(port);
	saddr.sin_addr.s_addr = inet_addr(ip);

	if (connect(sockfd, (struct sockaddr *)&saddr, sizeof(saddr)) == -1) {
		fprintf(stderr, "Unable to connect to %s:%u: %s\n", ip, port, strerror(errno));
		return -1;
	}

	return sockfd;
}

void directionCheckup(int ep, struct direction *dir) {
	unsigned int currentConnections = hashGetActive(dir->sessions);
	unsigned int create = 0;
	unsigned int delete = 0;

	if (currentConnections > dir->maxConnections) {
		delete = currentConnections - dir->maxConnections;
	} else if (currentConnections < dir->maxConnections) {
		create = dir->maxConnections - currentConnections;
	}

	if (debug) fprintf(stderr, "Connections - current: %u, required: %u, delete: %u, create: %u\n",
		currentConnections, dir->maxConnections, delete, create);

	// first delete
	unsigned int allCount;
	struct connection **allCons = (struct connection **)hashGetAll(dir->sessions, &allCount);

	if (allCount < delete) {
		fprintf(stderr, "Want to delete %u sessions but there are only %u in direction hash table",
			delete, allCount);
		delete = allCount;
	}

	unsigned int deleted = 0;
	for (unsigned int i = 0; i < delete; i++) {
		if (debug) fprintf(stderr, "Deleting connection %s:%u->%s:%u\n",
			allCons[i]->localIp, allCons[i]->localPort, allCons[i]->remoteIp, allCons[i]->remotePort);

		closeConnection(ep, allCons[i]);
		deleted++;
	}

	free(allCons);

	// then create new ones
	unsigned int created = 0;
	for (unsigned int i = 0; i < create; i++) {
		int sockfd = createConnection(dir->remoteIp, dir->remotePort);
		if (sockfd == -1) continue;

		// get some information about this socket 
		struct sockaddr_in laddr, raddr;
		socklen_t addrLen;

		addrLen = sizeof(laddr);
		if (getsockname(sockfd, (struct sockaddr *)&laddr, &addrLen) == -1) {
			fprintf(stderr, "Unable to get local socket information for direction %s:%u: %s",
				dir->remoteIp, dir->remotePort, strerror(errno));
			exit(3);
		}

		addrLen = sizeof(raddr);
		if (getpeername(sockfd, (struct sockaddr *)&raddr, &addrLen) == -1) {
			fprintf(stderr, "Unable to get remote socket information for direction %s:%u: %s",
				dir->remoteIp, dir->remotePort, strerror(errno));
			exit(3);
		}

		// and save them
		struct connection *con = (struct connection *)malloc(sizeof(struct connection));
		con->sockfd       = sockfd;
		con->direction    = dir;
		con->localIp      = strdup(inet_ntoa(laddr.sin_addr));
		con->localPort    = ntohs(laddr.sin_port);
		con->remoteIp     = strdup(inet_ntoa(raddr.sin_addr));
		con->remotePort   = ntohs(raddr.sin_port);
		time(&con->connected);
		con->lastSent     = con->connected;
		con->lastReceived = con->connected;
		con->pingRandom   = rand() % dir->pingTime;
		hashAdd(dir->sessions, con->sockfd, con);

		// add to epoll
		struct epoll_event event;
		event.data.ptr = con;
		event.events   = EPOLLIN; 

		if (epoll_ctl(ep, EPOLL_CTL_ADD, con->sockfd, &event) == -1) {
			fprintf(stderr, "Unable to add client %s:%u->%s:%u to epoll: %s\n",
				con->localIp, con->localPort, con->remoteIp, con->remotePort,
				strerror(errno));
			exit(3);
		}

		created++;
		if (debug) fprintf(stderr, "Created connection %s:%u->%s:%u\n",
			con->localIp, con->localPort, con->remoteIp, con->remotePort);
	}

	if (debug) fprintf(stderr, "For direction %s:%u in this cycle deleted %u and created %u connections\n",
		dir->remoteIp, dir->remotePort, deleted, created);
}

void closeConnection(int ep, struct connection *con) {
	if (epoll_ctl(ep, EPOLL_CTL_DEL, con->sockfd, NULL) == -1) {
		fprintf(stderr, "Unable to remove client %s:%u->%s:%u from epoll: %s\n",
			con->localIp, con->localPort, con->remoteIp, con->remotePort,
			strerror(errno));
		exit(4);
	}

	shutdown(con->sockfd, SHUT_RDWR);
	close(con->sockfd);

	hashDelete(con->direction->sessions, con->sockfd);

	free(con->localIp);
	free(con->remoteIp);
	free(con);
}

void closeAllConnections(int ep, struct direction *dir) {
		unsigned int allCount;
		struct connection **allCons = (struct connection **)hashGetAll(dir->sessions, &allCount);

		for (unsigned int i = 0; i < allCount; i++) {
			if (debug) fprintf(stderr, "Closing connection %s:%d->%s:%d\n",
				allCons[i]->localIp, allCons[i]->localPort,
				allCons[i]->remoteIp, allCons[i]->remotePort);

			closeConnection(ep, allCons[i]);
		}

		free(allCons);
}


void directionPing(int ep, struct direction *dir) {
	unsigned int allCount;
	struct connection **allCons = (struct connection **)hashGetAll(dir->sessions, &allCount);
	unsigned int count = 0;

	time_t now;
	time(&now);

	char ping[1024];
	snprintf(ping, 1024-1, "PING");

	for (unsigned int i = 0; i < allCount; i++) {
		if ((now-allCons[i]->lastSent) > (dir->pingTime + allCons[i]->pingRandom)) {

			if (send(allCons[i]->sockfd, ping, strlen(ping), 0) != strlen(ping)) {
				fprintf(stderr, "Cannot ping client %s:%u->%s:%u: %s\n",
					allCons[i]->localIp, allCons[i]->localPort, allCons[i]->remoteIp, allCons[i]->remotePort,
					strerror(errno));
			} else {
				if (debug) fprintf(stderr, "Pinged client %s:%u->%s:%u\n",
					allCons[i]->localIp, allCons[i]->localPort, allCons[i]->remoteIp, allCons[i]->remotePort);

				time(&allCons[i]->lastSent);
				count++;

				// do not spend to much CPU time with ping in one cycle - finish next time
				if (count > dir->pingMax) break;
			}
		}
	}

	free(allCons);

	if (debug) fprintf(stderr, "For direction %s:%u in this cycle pinged %u connections\n",
		dir->remoteIp, dir->remotePort, count);
}

void handleEpoll(int ep) {
	struct epoll_event events[32768];

	for (;;) {
		if (terminate) { break; }

		int nfds = epoll_wait(ep, events, 32768, 250);
		if (nfds == -1) {
			fprintf(stderr, "Cannot wait for epoll events: %s\n", strerror(errno));
			exit(4);
		}
	
		for (int i = 0; i < nfds; i++) {
			struct connection *con = (struct connection *)events[i].data.ptr;
			char buffer[1024];
	
			if ((events[i].events & EPOLLIN) == EPOLLIN) {
				int r = recv(con->sockfd, buffer, 1024-1, 0);
				if (r == -1) {
					fprintf(stderr, "Error on connection %s:%u->%s:%u: %s\n",
						con->localIp, con->localPort, con->remoteIp, con->remotePort,
						strerror(errno));
	
					closeConnection(ep, con);
					continue;
	
				} else if (r == 0) {
					if (debug) fprintf(stderr, "Connection %s:%u->%s:%u closed by remote side\n",
						con->localIp, con->localPort, con->remoteIp, con->remotePort);
	
					closeConnection(ep, con);
					continue;
				}
	
				// we have got some data
				buffer[r] = 0;
				time(&con->lastReceived);
				if (debug) fprintf(stderr, "Incoming data from %s:%u->%s:%u: %s\n",
						con->localIp, con->localPort, con->remoteIp, con->remotePort,
						buffer);
	
			} else {
				fprintf(stderr, "Unknown event from client socket: 0x%x\n", events[i].events);
				exit(4);
			}
		}
	}
}

struct workerParam {
	int               ep;
	struct direction  **directions;
	unsigned int      directionCount;
};

void *worker1(void *param) {
	struct workerParam *wp = (struct workerParam *)param;
	while (!terminate) {
		for (int i = 0; i < wp->directionCount; i++) {
			directionCheckup(wp->ep, wp->directions[i]);
			if (ping) directionPing(wp->ep, wp->directions[i]);
		}	
		sleep(1);
	}
}

void *worker2(void *param) {
	struct workerParam *wp = (struct workerParam *)param;
	handleEpoll(wp->ep);
}

int main(int argc, char *argv[]) {
	//
	struct workerParam wp;
	wp.ep              = epoll_create(100000);
	wp.directions      = NULL;
	wp.directionCount  = 0;

	int ulimit_fd = 1;

	while (1) {
		static struct option long_options[] = {
			{"target",     required_argument, 0,  't' },
			{"hashsize",   required_argument, 0,  's' },
			{"debug",      no_argument,       0,  'd' },
			{"ping",       no_argument,       0,  'p' },
			{"help",       no_argument,       0,  'h' },
			{"no-ulimit",  no_argument,       0,  'u' },
		};

		int c = getopt_long(argc, argv, "t:s:dphu", long_options, NULL);
		if (c == -1) break;
		
		switch (c) {
		case 't':;
			// first split by colon
			char *colon = strchr(optarg, ':');
			if (colon == NULL) {
				fprintf(stderr, "Error: option --connect parameter must be in the format of \"ip:port\" or \"ip:port-port\"!\n");
				exit(1);
			}

			char *lAddr = strndup(optarg, colon-optarg);
			unsigned short lPortStart, lPortEnd;

			// check if it is a range
			char *dash = strchr(colon+1, '-');
			if (dash == NULL) {
				lPortStart = atoi(colon+1);
				lPortEnd   = lPortStart;
			} else {
				lPortEnd   = atoi(dash+1);
				dash[0] = 0;
				lPortStart = atoi(colon+1);
			}

			if (lPortStart > lPortEnd) {
				fprintf(stderr, "Error: end port must not be lower than start port!\n");
				exit(1);
			}

			if (debug) fprintf(stderr, "Preparing direction to %s with remote ports %d-%d\n", lAddr, lPortStart, lPortEnd);
			for (unsigned short port = lPortStart; port <= lPortEnd; port++) {
				wp.directions = realloc(wp.directions, sizeof(struct direction **)*(wp.directionCount+1));
				wp.directions[wp.directionCount] = directionInit(lAddr, port, 1);
				wp.directionCount++;
			}
			free(lAddr);

			break;
		case 's':
			hashSize = atoi(optarg);
			break;

		case 'd':
			debug = 1;
			break;

		case 'u':
			ulimit_fd = 0;
			break;

		case 'p':
			ping = 1;
			break;

		case 'h':
			fprintf(stderr, "Usage: %s [-dph] --target ip:port[-port] [--hashsize NUM]\n", argv[0]);
			fprintf(stderr, "       -h         ... show this help\n");
			fprintf(stderr, "       -d         ... enable debug outputs\n");
			fprintf(stderr, "       -p         ... ping through connections periodically\n");
			fprintf(stderr, "       -u         ... do not try to change resource limits on file descriptors\n");
			fprintf(stderr, "       --target   ... connect on IP address and port\n");
			fprintf(stderr, "                      optionally port range can be specified\n");
			fprintf(stderr, "                      this option can appear multiple times\n");
			fprintf(stderr, "       --hashsize ... size of primary hashing table for following targets\n");
			return 0;
			break;
		}
	}

	if (wp.directionCount == 0) {
		fprintf(stderr, "No remote ends specified.\nUse --target parameter to specify at least one or -h for more information.\n");
		exit(1);
	}

	if (ulimit_fd) {
		// disable fd limits
		FILE *nropen = fopen("/proc/sys/fs/nr_open", "r");
		if (nropen == NULL) {
			fprintf(stderr, "Unable to disable descriptors limit: %s\n", strerror(errno));
		}
	
		unsigned int maxfiles;
		if (fscanf(nropen, "%u", &maxfiles) != 1) {
			fprintf(stderr, "Unable to disable descriptors limit: /proc/sys/fs/nr_open format error\n");
		}
		fclose(nropen);
	
		struct rlimit limits;
		limits.rlim_cur = maxfiles;
		limits.rlim_max = maxfiles;
	
		if (setrlimit(RLIMIT_NOFILE, &limits) == -1) {
			fprintf(stderr, "Unable to disable descriptors limit: %s\n", strerror(errno));
		}
	}
	
	struct rlimit limits;
	if (getrlimit(RLIMIT_NOFILE, &limits) == -1) {
		fprintf(stderr, "Unable to get descriptors limit: %s\n", strerror(errno));
	}

	pthread_t workerThread1, workerThread2;
	if (pthread_create(&workerThread1, NULL, worker1, &wp) == -1) {
		fprintf(stderr, "Unable to start worker thread1: %s\n", strerror(errno));
		exit(5);
	}
	if (pthread_create(&workerThread2, NULL, worker2, &wp) == -1) {
		fprintf(stderr, "Unable to start worker thread2: %s\n", strerror(errno));
		exit(5);
	}

	char last[1024];
	bzero(last, 1024);

	for (;;) {
		fprintf(stdout, "initiator $ ");

		char line[1024];
		if (fgets(line, 1024, stdin) == NULL) {
			fprintf(stdout, "\n");
			strcpy(line, "quit\n");
		}
		while (line[strlen(line)-1] == '\n') { line[strlen(line)-1] = 0; }

		if (strlen(line) == 0) {
			strcpy(line, last);
		}

		if (strcmp(line, "status") == 0) {
			unsigned int currentConnections = 0;
			unsigned int currentRequest     = 0;
			for (unsigned int i = 0; i < wp.directionCount; i++) {
				currentConnections += hashGetActive(wp.directions[i]->sessions);
				currentRequest     += wp.directions[i]->maxConnections;
			}

			fprintf(stdout, "Connections requested    : %u\n", currentRequest);
			fprintf(stdout, "Connected clients        : %u\n", currentConnections);
			fprintf(stdout, "Configured targets       : %u\n", wp.directionCount);
			fprintf(stdout, "Max descriptors limit    : %u/%u\n", limits.rlim_cur, limits.rlim_max);
			if (ping) 
				fprintf(stdout, "Ping through connections : yes\n");
			else
				fprintf(stdout, "Ping through connections : no\n");
			if (debug) 
				fprintf(stdout, "Debug output             : yes\n");
			else
				fprintf(stdout, "Debug output             : no\n");

		} else if (strcmp(line, "details") == 0) {
			for (unsigned int i = 0; i < wp.directionCount; i++) {
				fprintf(stdout, "Target %s:%u:\n", wp.directions[i]->remoteIp, wp.directions[i]->remotePort);
				fprintf(stdout, "   Maximum allowed connections       : %u\n", wp.directions[i]->maxConnections);
				fprintf(stdout, "   Currently established connections : %u\n", hashGetActive(wp.directions[i]->sessions));
				fprintf(stdout, "   Session hash max depth            : %u\n", hashGetDepth(wp.directions[i]->sessions));
			}

		} else if (strncmp(line, "set", 3) == 0) {
				if (strlen(line) <= 4) {	
					fprintf(stdout, "You need to specify the number to requested connections!\n");
				} else {
					unsigned int request = atoi(line+4);
					if (debug) fprintf(stderr, "Setting total number of requested connections to %u\n", request);

					unsigned int requestForOne = request / wp.directionCount;
					for (unsigned int i = 0; i < wp.directionCount; i++) {
						if (i == 0)
							wp.directions[i]->maxConnections = request - ((wp.directionCount-1) * requestForOne);
						else
							wp.directions[i]->maxConnections = requestForOne;
					}
				}

		} else if (strcmp(line, "debug") == 0) {
			if (debug == 0) {
				debug = 1;
				fprintf(stdout, "Debug output enabed\n");
			} else {
				debug = 0;
				fprintf(stdout, "Debug output disabled\n");
			}

		} else if (strcmp(line, "ping") == 0) {
			if (ping == 0) {
				ping = 1;
				fprintf(stdout, "Ping enabed\n");
			} else {
				ping = 0;
				fprintf(stdout, "Ping disabled\n");
			}

		} else if (strcmp(line, "quit") == 0) {
			terminate = 1;
			pthread_join(workerThread1, NULL);
			pthread_join(workerThread2, NULL);
			break;

		} else {
			fprintf(stdout, "Unknown command \"%s\", valid commands are:\n", line);
			fprintf(stdout, "- status          ... show number of currently connected clients\n");
			fprintf(stdout, "- details         ... show more information about each target\n");
			fprintf(stdout, "- set NUM         ... set the number of required connections\n");
			fprintf(stdout, "- debug           ... toggle the debug log\n");
			fprintf(stdout, "- ping            ... toggle the ping through the established connections\n");
			fprintf(stdout, "- quit            ... close all clients and exit\n");
			continue;
		}

		strcpy(last, line);
	}

	for (unsigned int i = 0; i < wp.directionCount; i++) {
		closeAllConnections(wp.ep, wp.directions[i]);
		hashDestroy(wp.directions[i]->sessions);
		free(wp.directions[i]->remoteIp);
		free(wp.directions[i]);
	}
	close(wp.ep);
	free(wp.directions);
}
